torchmetrics/average.py

```diff
@@ -0,0 +1,101 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Callable, Optional
+
+import torch
+from torch import Tensor
+
+from torchmetrics.metric import Metric
+
+
+class AverageMeter(Metric):
+    """Computes the average of a stream of values.
+
+    Forward accepts
+        - ``value`` (float tensor): ``(...)``
+        - ``weight`` (float tensor): ``(...)``
+
+    Args:
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is
+            set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step.
+        process_group:
+            Specify the process group on which synchronization is called.
+            default: None (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state.
+            When `None`, DDP will be used to perform the allgather.
+
+    Example::
+        >>> from torchmetrics import AverageMeter
+        >>> avg = AverageMeter()
+        >>> avg.update(3)
+        >>> avg.update(1)
+        >>> avg.compute()
+        tensor(2.)
+
+        >>> avg = AverageMeter()
+        >>> values = torch.tensor([1., 2., 3.])
+        >>> avg(values)
+        tensor(2.)
+
+        >>> avg = AverageMeter()
+        >>> values = torch.tensor([1., 2.])
+        >>> weights = torch.tensor([3., 1.])
+        >>> avg(values, weights)
+        tensor(1.2500)
+    """
+
+    def __init__(
+        self,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+        self.add_state("value", torch.zeros(()), dist_reduce_fx="sum")
+        self.add_state("weight", torch.zeros(()), dist_reduce_fx="sum")
+
+    # TODO: these should be Union[Tensor, float], but Unions are not picklable
+    # in Python 3.6. https://github.com/python/cpython/pull/6216 fixed this in
+    # Python 3.7+
+    def update(self, value: Tensor, weight: Tensor = 1.0) -> None:
```

Comment:
I'm not sure how you want to track this, so I just left it as a code comment for now.

Reply:
Why is pickleability an issue for type annotations? You can probably do `from __future__ import annotations` and they become strings.

Reply:
Otherwise you could probably make them strings manually (not shure how this works out with nested typing though)

Reply:
I think strings manually works here: `from __future__ import annotations` doesn't exist yet on Python 3.6